externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLLineageParseHelper.scala

```diff
@@ -0,0 +1,366 @@
+              getCurrentPlanField[CatalogPlugin](
+                getPlanMethod[LogicalPlan]("left"),
+                "catalog").name())
+          }
+        extractColumnsLineage(getQuery, parentColumnsLineage).map { case (k, v) =>
+          s"$catalog.$namespace.$table.$k" -> v
+        }
+
+      case p if p.nodeName == "InsertIntoDataSourceCommand" =>
+        val logicalRelation = getPlanField[LogicalRelation]("logicalRelation")
+        val table = logicalRelation.catalogTable.map(_.qualifiedName).getOrElse("")
+        extractColumnsLineage(getQuery, parentColumnsLineage).map { case (k, v) =>
+          s"$table.$k" -> v
+        }
+
+      case p if p.nodeName == "InsertIntoHadoopFsRelationCommand" =>
+        val table =
+          getPlanField[Option[CatalogTable]]("catalogTable").map(_.qualifiedName).getOrElse("")
+        extractColumnsLineage(getQuery, parentColumnsLineage).map { case (k, v) =>
+          s"$table.$k" -> v
+        }
+
+      case p
+          if p.nodeName == "InsertIntoDataSourceDirCommand" ||
+            p.nodeName == "InsertIntoHiveDirCommand" =>
+        val dir =
+          getPlanField[CatalogStorageFormat]("storage").locationUri.map(_.toString).getOrElse("")
+        extractColumnsLineage(getQuery, parentColumnsLineage).map { case (k, v) =>
+          s"`$dir`.$k" -> v
+        }
+
+      case p if p.nodeName == "InsertIntoHiveTable" =>
+        val table = getPlanField[CatalogTable]("table").qualifiedName
+        extractColumnsLineage(getQuery, parentColumnsLineage).map { case (k, v) =>
+          s"$table.$k" -> v
+        }
+
+      case p if p.nodeName == "SaveIntoDataSourceCommand" =>
+        // TODO: Should get the table via datasource options?
+        extractColumnsLineage(getQuery, parentColumnsLineage)
+
+      // For query
+      case p: Project =>
+        val nextColumnsLineage =
+          joinColumnsLineage(parentColumnsLineage, getSelectColumnLineage(p.projectList))
+        p.children.map(extractColumnsLineage(_, nextColumnsLineage)).reduce(mergeColumnsLineage)
+
+      case p: Aggregate =>
+        val nextColumnsLineage =
+          joinColumnsLineage(parentColumnsLineage, getSelectColumnLineage(p.aggregateExpressions))
+        p.children.map(extractColumnsLineage(_, nextColumnsLineage)).reduce(mergeColumnsLineage)
+
+      case p: Union =>
+        // merge all children in to one derivedColumns
+        val childrenUnion =
+          p.children.map(extractColumnsLineage(_, ListMap())).map(_.values).reduce {
+            (left, right) =>
+              left.zip(right).map(col => (col._1 ++ col._2).distinct)
+          }
+        val childrenColumnsLineage =
+          ListMap(p.output.map(a => s"${a.name}#${a.exprId.id}").toList.zip(childrenUnion): _*)
+        joinColumnsLineage(parentColumnsLineage, childrenColumnsLineage)
+
+      case p: SubqueryAlias =>
+        extractColumnsLineage(p.child, parentColumnsLineage)
```

Comment:
it seems never reach here with optimized plan

Reply:
Yes, at first it was analyzed plan, but finally it was changed to optmized plan. we can remove it.